clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java

```diff
@@ -0,0 +1,3630 @@
+import static org.apache.kafka.test.TestUtils.assertOptional;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class FetchRequestManagerTest {
+
+    private static final double EPSILON = 0.0001;
+
+    private ConsumerRebalanceListener listener = new NoOpConsumerRebalanceListener();
+    private String topicName = "test";
+    private String groupId = "test-group";
+    private Uuid topicId = Uuid.randomUuid();
+    private Map<String, Uuid> topicIds = new HashMap<String, Uuid>() {
+        {
+            put(topicName, topicId);
+        }
+    };
+    private Map<Uuid, String> topicNames = singletonMap(topicId, topicName);
+    private final String metricGroup = "consumer" + groupId + "-fetch-manager-metrics";
+    private TopicPartition tp0 = new TopicPartition(topicName, 0);
+    private TopicPartition tp1 = new TopicPartition(topicName, 1);
+    private TopicPartition tp2 = new TopicPartition(topicName, 2);
+    private TopicPartition tp3 = new TopicPartition(topicName, 3);
+    private TopicIdPartition tidp0 = new TopicIdPartition(topicId, tp0);
+    private TopicIdPartition tidp1 = new TopicIdPartition(topicId, tp1);
+    private TopicIdPartition tidp2 = new TopicIdPartition(topicId, tp2);
+    private TopicIdPartition tidp3 = new TopicIdPartition(topicId, tp3);
+    private int validLeaderEpoch = 0;
+    private MetadataResponse initialUpdateResponse =
+        RequestTestUtils.metadataUpdateWithIds(1, singletonMap(topicName, 4), topicIds);
+
+    private int minBytes = 1;
+    private int maxBytes = Integer.MAX_VALUE;
+    private int maxWaitMs = 0;
+    private int fetchSize = 1000;
+    private long retryBackoffMs = 100;
+    private int requestTimeoutMs = 30000;
+    private MockTime time = new MockTime(1);
+    private SubscriptionState subscriptions;
+    private ConsumerMetadata metadata;
+    private FetchMetricsRegistry metricsRegistry;
+    private FetchMetricsManager metricsManager;
+    private MockClient client;
+    private Metrics metrics;
+    private final ApiVersions apiVersions = new ApiVersions();
+    private ConsumerNetworkClient consumerClient;
+    private FetchRequestManager<byte[], byte[]> fetchRequestManager;
+    private FetchCollector<byte[], byte[]> fetchCollector;
+    private TestableNetworkClientDelegate networkClientDelegate;
```

Comment:
I'm still trying to fully understand the specifics of the fetch manager related to this, but after a first pass I wonder why we need/want to involve a network client for unit testing the fetch manager... We should aim for a separation of the components, and test here the fetch manager logic without involving the network client. We've been doing it like that for the other managers, where we have been able to test the managers logic for building the requests, and then for processing the client response received (abstracting all the magic that the network client would do under the covers)

Reply:
@lianetm You're right, the `FetchRequestManagerTest` is unique among the request managers in that it uses the `NetworkClientDelegate`. The other request manager test classes rely on `mock`-ing and/or manually creating the `ClientResponse` to determine correctness. That's definitely a useful approach, but the current approach for the `FetchRequestManagerTest` has merit for the following reasons:

1. It doesn't "rock the boat"—`FetcherTest` uses the `ConsumerNetworkClient` throughout the tests and `FetchRequestManagerTest` is lifted from `FetcherTest` so it inherits those tests. The unit tests from `FetchTest` are exhaustive in their coverage, so I'm reluctant to rework them to the point where the use of the `NetworkClientDelegate` could be removed. (Note that the `NetworkClientDelegate` here wraps a `MockClient`, so it's not exercising the low-level network path.) In addition, the goal—once the old `Fetcher` is removed—is to replace `FetcherTest` with `FetchRequestManagerTest`. So keeping the code in the new test as close to the original as reasonably possible ensures the test coverage is comparable and hopefully makes the merge process less onerous.
2. `mock` can mask bugs—During development, I tried to integrate my code into `DefaultBackgroundThreadTest`, `DefaultEventHandlerTest`, and `PrototypeAsyncConsumerTest`. However, these tests were so heavily `mock`-ed that they neglected to exercise basic functionality of their respective companion classes. (This lack of coverage was borne out by code coverage tool results.) I uncovered a painful amount of silly errors in the underlying code, which took a while to resolve.
3. Fetching logic is obstinately intertwined and spread across multiple classes. It could be argued that the `FetchRequestManagerTest` is closer to an integration test than a true unit test. The fetch process is pretty complicated and spans `AbstractFetch`, `CompletedFetch`, `Fetch`, `FetchBuffer`, `FetchCollector`, `FetchRequestManager`, and maybe even a couple of more. Without incorporating most of those components in the tests, the resulting tests would be anemically inferior to what was in `FetchTest`.

I had originally thought to have a second `FetchRequestManagerTest` that was simpler and relied on `mock`s, but ultimately I felt that the tests provided by the `FetcherTest` covered those anyway 🤷‍♂️

Thoughts?

Reply:
I totally see how we are kind of inhering this, but my point is that it looks like the networkClient is mainly involved for preparing a client response, and processing it with poll (which will ultimately pass it to the manager that we want to test), so I'm trying to see the value it brings passing all requests/responses through that network client here. If we just poll the manager to build the requests (no network client involved), and then instead of:
`client.prepareResponse(fullFetchResponse(tidp0, this.records, Errors.NONE, 100L, 0));
networkClientDelegate.poll(time.timer(0));` 
We just provide the manager with the client response like:
`NetworkClientDelegate.PollResult pollResult = fetchRequestManager.poll(time.milliseconds());
ClientResponse clientResponse = new ClientResponse(
        new RequestHeader(ApiKeys.FETCH, (short) 1, "", 1),
        ...
        fetchResponse);
pollResult.unsentRequests.get(0).future().complete(clientResponse);`
With the above change we would be testing the manager functionality that we want, no need to involve a network client component. 
As an exercise I tried these changes in a couple of tests

Reply:
I've started working to convert to this style now.

Reply:
OK, I can't make the transition to the simplified approach 😢

Around 85% of the methods can use the suggested approach, but there are about a dozen tests for which it won't work. Those remaining tests are exercising corner case behavior related to the interplay between the client and the fetcher for things like disconnects, timeouts, metadata changes, wakeup, etc. Any attempts I have made require a good amount of spaghetti code scaffolding that ends up duplicating functionality from `MockClient` anyway 🤷‍♂️ 

'tis no exaggeration—I have spent around 60 hours trying to reuse `FetcherTest` for `FetchRequestManager`. I was able to get the tests working nearly a month ago, and since then have been attempting to refactor to make things simpler and reduce diffs between my changes and `ctr-staging` and AK `trunk`.

At this point, the fetch request manager code is very well tested on account of leveraging `FetcherTest`. I would like to propose that we make a note of the need for future refactoring, use the `FetchRequestManagerTest` in its current, imperfect form, and move on.

WDYT?